stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -147,6 +147,89 @@ const ClassMetadata *TypeReference::getObjCClass(TypeReferenceKind kind) const {
 }
 #endif
 
+static MetadataState
+tryGetCompleteMetadataNonblocking(const Metadata *metadata) {
+  return swift_checkMetadataState(
+             MetadataRequest(MetadataState::Complete, /*isNonBlocking*/ true),
+             metadata)
+      .State;
+}
+
+/// Get the superclass of metadata, which may be incomplete. When the metadata
+/// is not sufficiently complete, then we fall back to demangling the superclass
+/// in the nominal type descriptor, which is slow but works. Return NULL if the
+/// metadata is not a class.
+///
+/// If the metadata's current state is known, it may be passed in as
+/// knownMetadataState. This saves the cost of retrieving that info separately.
+static MetadataResponse getSuperclassForMaybeIncompleteMetadata(
+    const Metadata *metadata,
+    llvm::Optional<MetadataState> knownMetadataState) {
+  const ClassMetadata *classMetadata = dyn_cast<ClassMetadata>(metadata);
+  if (!classMetadata)
+    return {_swift_class_getSuperclass(metadata), MetadataState::Complete};
+
+  MetadataState metadataState;
+  if (knownMetadataState)
+    metadataState = *knownMetadataState;
+  else
+    metadataState = tryGetCompleteMetadataNonblocking(classMetadata);
+
+  if (metadataState == MetadataState::Complete) {
+    // The subclass metadata is complete. Fetch and return the superclass.
+    return {_swift_class_getSuperclass(metadata), MetadataState::Complete};
```

Comment:
Can this just be `classMetadata->Superclass`?

Reply:
That isn't sufficient because this code now potentially goes into the ObjC hierarchy. However, I should just be using `getMetadataForClass` on the field, rather than calling out to `_swift_class_getSuperclass` which does much more than we need here. I updated the PR to use that instead. Thanks!